src/cpp/src/config.cpp

```diff
@@ -4,246 +4,392 @@
+                        }
+                        for (OptInfo<float> v : opt_info.f_var_map) {
+                            if (section == v.s_section && option_name == v.s_option) {
+                                *(v.var_ptr) = std::stof(value);
+                                valid = true;
+                            }
+                        }
+                        for (OptInfo<bool> v : opt_info.b_var_map) {
+                            if (section == v.s_section && option_name == v.s_option) {
+                                *(v.var_ptr) = (value == "true");
+                                valid = true;
+                            }
+                        }
+                        if (!valid) {
+                            throw std::exception();
+                        }
+                    } else {
+                        throw std::exception();
+                    }
+                }
+            } catch (std::exception) {
+                throw cxxopts::option_syntax_exception("Unable to parse supplied command line configuration options");
+            }
+        }
+    } catch (const cxxopts::OptionException& e) {
+        SPDLOG_ERROR("Error parsing options: {}", e.what());
+        std::cout << cmd_options.help() << std::endl;
+        exit(-1);
+    }
+}
+
+void validateNumericalOptions(AllOptInfo opt_info) {
+    for (OptInfo<int64_t> v : opt_info.i64_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
     }
+    for (OptInfo<int> v : opt_info.i_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
+    }
+    for (OptInfo<float> v : opt_info.f_var_map) {
+        if (*(v.var_ptr) < v.range[0] || *(v.var_ptr) > v.range[1]) {
+            SPDLOG_ERROR("{}.{}: value {} out of range [{}, {}]", v.s_section, v.s_option, *(v.var_ptr), v.range[0], v.range[1]);
+            exit(-1);
+        }
+    }
+}
+
+MariusOptions parseConfig(int64_t argc, char *argv[]) {
+
+    // Associate each option with:
+    // variable, default value, section name, user option name, (optional) range
+
+    // List for each type of variable
+    std::vector<OptInfo<std::string>> s_var_map;
+    std::vector<OptInfo<int64_t>> i64_var_map;
+    std::vector<OptInfo<int>> i_var_map;
+    std::vector<OptInfo<float>> f_var_map;
+    std::vector<OptInfo<bool>> b_var_map;
+
+    float FLOAT_MAX = std::numeric_limits<float>::max();
 
     // General options
-    torch::DeviceType device;
+    torch::DeviceType device; // Device to use for training
     string s_device;
+    s_var_map.push_back((OptInfo<std::string>){&s_device, "CPU", "general", "device"});
```

Comment:
Can we put the variable map stuff into a helper function? Will make parseConfig a bit easier to follow.